internal/oci/oci_test.go

```diff
@@ -0,0 +1,173 @@
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/k0sproject/k0s/internal/oci"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"sigs.k8s.io/yaml"
+)
+
+//go:embed testdata/*
+var testData embed.FS
+
+// we define our tests as yaml files inside the testdata directory. this
+// function parses them and returns a map of the tests.
+func parseTestsYAML[T any](t *testing.T) map[string]T {
+	entries, err := testData.ReadDir("testdata")
+	require.NoError(t, err)
+	tests := make(map[string]T, 0)
+	for _, entry := range entries {
+		fpath := path.Join("testdata", entry.Name())
+		data, err := testData.ReadFile(fpath)
+		require.NoError(t, err)
+
+		var onetest T
+		err = yaml.Unmarshal(data, &onetest)
+		require.NoError(t, err)
+
+		tests[fpath] = onetest
+	}
+	return tests
+}
+
+// testFile represents a single test file inside the testdata directory.
+type testFile struct {
+	Manifest      string            `json:"manifest"`
+	Expected      string            `json:"expected"`
+	Error         string            `json:"error"`
+	Authenticated bool              `json:"authenticated"`
+	AuthUser      string            `json:"authUser"`
+	AuthPass      string            `json:"authPass"`
+	Artifacts     map[string]string `json:"artifacts"`
+}
+
+func TestDownload(t *testing.T) {
+	for tname, tt := range parseTestsYAML[testFile](t) {
+		t.Run(tname, func(t *testing.T) {
+			addr := startOCIMockServer(t, tname, tt)
+
+			opts := []oci.DownloadOption{oci.WithInsecureSkipTLSVerify()}
+			if tt.Authenticated {
+				entry := oci.DockerConfigEntry{tt.AuthUser, tt.AuthPass}
+				opts = append(opts, oci.WithDockerAuth(
+					oci.DockerConfig{
+						Auths: map[string]oci.DockerConfigEntry{
+							addr: entry,
+						},
+					},
+				))
+			}
+
+			buf := bytes.NewBuffer(nil)
+			url := fmt.Sprintf("%s/repository/artifact:latest", addr)
+			err := oci.Download(context.TODO(), url, buf, opts...)
+			if tt.Expected != "" {
+				require.NoError(t, err)
+				require.Empty(t, tt.Error)
+				require.Equal(t, tt.Expected, buf.String())
+				return
+			}
+			require.NotEmpty(t, tt.Error)
+			require.ErrorContains(t, err, tt.Error)
+		})
+	}
+}
+
+// startOCIMockServer starts a mock server that will respond to the given test.
+// this mimics the behavior of the real OCI registry. This function returns the
+// address of the server.
+func startOCIMockServer(t *testing.T, tname string, test testFile) string {
+	var serverAddr string
+	server := httptest.NewTLSServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+			// this is a request to authenticate.
+			if strings.Contains(r.URL.Path, "/token") {
+				user, pass, _ := r.BasicAuth()
+				if user != "user" || pass != "pass" {
+					w.WriteHeader(http.StatusUnauthorized)
+					return
+				}
+				res := map[string]string{"token": tname}
+				marshalled, err := json.Marshal(res)
+				require.NoError(t, err)
```

Comment:
Better use `assert` here, as `require` will call `t.FailNow()` but we're not in the right goroutine for that. Same for the other `require` usages below.